executor/src/witgen/rows.rs

```diff
@@ -16,6 +20,85 @@ use super::{
     FixedData,
 };
 
+/// A small wrapper around a row index, which knows the total number of rows.
+/// When converted to DegreeType or usize, it will be reduced modulo the number of rows
+/// (handling negative indices as well).
+#[derive(Clone, Copy)]
+pub struct RowIndex {
+    index: i64,
+    num_rows: DegreeType,
+}
+
+impl From<RowIndex> for DegreeType {
+    fn from(row_index: RowIndex) -> Self {
+        // Ensure that 0 <= index < num_rows
+        if row_index.index >= 0 {
+            (row_index.index as DegreeType) % row_index.num_rows
+        } else {
+            row_index.num_rows - ((-row_index.index as DegreeType) % row_index.num_rows)
```

Comment:
Isn't this wrong for `row_index.index == -row_index.num_rows`?
Maybe we should instead assume that it is usually at most `+- 2*num_rows` and thus instead add/subtract num_rows until we are in range?

Reply:
Right, good catch! I think I fixed this by rearranging the parentheses!

Reply:
OK, never mind, didn't work. I just added an assertion